File: src/main/java/com/lessmarkup/framework/helpers/JsonSerializer.java
package com.lessmarkup.framework.helpers;

import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonPrimitive;
import com.lessmarkup.interfaces.exceptions.CommonException;
import java.time.OffsetDateTime;
import java.util.OptionalLong;

public final class JsonSerializer {
    
    public static <T> T deserializePojo(Class<T> type, String data) {
        return deserializePojo(type, deserializeToTree(data));
    }
    
    public static JsonElement serializePojoToTree(Object object) {
        
        Class<?> type = object.getClass();
        if (type.equals(long.class)) {
            return new JsonPrimitive((long) object);
        } else if (type.equals(int.class)) {
            return new JsonPrimitive((int) object);
        } else if (type.equals(String.class)) {
            return new JsonPrimitive((String) object);
        } else if (JsonElement.class.isAssignableFrom(type)) {
            return (JsonElement) object;
        }
        
        JsonObject ret = new JsonObject();
        for (PropertyDescriptor property : TypeHelper.getProperties(object.getClass())) {
            if (property.getType().equals(long.class)) {
                ret.addProperty(property.getName(), (long) property.getValue(object));
            } else if (property.getType().equals(String.class)) {
                ret.addProperty(property.getName(), (String) property.getValue(object));
            } else if (property.getType().equals(boolean.class)) {
                ret.addProperty(property.getName(), (boolean) property.getValue(object));
            } else if (property.getType().equals(int.class)) {
                ret.addProperty(property.getName(), (int) property.getValue(object));
            } else if (property.getType().equals(OptionalLong.class)) {
                OptionalLong value = (OptionalLong) property.getValue(object);
                if (value.isPresent()) {
                    ret.addProperty(property.getName(), value.getAsLong());
                } else {
                    ret.add(property.getName(), JsonNull.INSTANCE);
                }
            } else if (property.getType().equals(OffsetDateTime.class)) {
                ret.addProperty(property.getName(), ((OffsetDateTime)property.getValue(object)).toString());
            } else if (JsonElement.class.isAssignableFrom(type)) {
                ret.add(property.getName(), (JsonElement) property.getValue(object));
            }
        }
        return ret;
    }
    
    public static String serializePojo(Object object) {
        JsonElement obj = serializePojoToTree(object);
        return obj.toString();
    }

    public static JsonElement deserializeToTree(String data) {
        return new JsonParser().parse(data);
    }

    @SuppressWarnings("unchecked")
    public static <T> T deserializePojo(Class<T> type, JsonElement data) {
        
        if (type.equals(int.class)) {
            return (T)(Integer) data.getAsInt();
        } else if (type.equals(long.class)) {
            return (T)(Long) data.getAsLong();
        } else if (type.equals(String.class)) {
            return (T) data.getAsString();
        } else if (type.equals(OptionalLong.class)) {
            if (data.isJsonNull()) {
                return (T) OptionalLong.empty();
            } else {
                return (T) OptionalLong.of(data.getAsLong());
            }
        } else if (JsonElement.class.isAssignableFrom(type)) {
            return (T) data;
        }
        
        if (data.isJsonNull() || !data.isJsonObject()) {
            return null;
        }
        
        JsonObject object = data.getAsJsonObject();
        
        T ret;
        
        try {
            ret = type.newInstance();
        } catch (InstantiationException | IllegalAccessException ex) {
            throw new CommonException(ex);
        }
        
        for (PropertyDescriptor property : TypeHelper.getProperties(type)) {
            JsonElement element = object.get(property.getName());
            if (element == null) {
                continue;
            }
            
            if (property.getType().equals(long.class)) {
                property.setValue(ret, element.getAsLong());
            } else if (property.getType().equals(String.class)) {
                property.setValue(ret, element.getAsString());
            } else if (property.getType().equals(boolean.class)) {
                property.setValue(ret, element.getAsBoolean());
            } else if (property.getType().equals(int.class)) {
                property.setValue(ret, element.getAsInt());
            } else if (property.getType().equals(OptionalLong.class)) {
                if (!element.isJsonNull()) {
                    property.setValue(ret, element.getAsLong());
                }
            } else if (property.getType().equals(OffsetDateTime.class)) {
                if (!element.isJsonNull()) {
                    property.setValue(ret, OffsetDateTime.parse(element.getAsString()));
                }
            } else if (JsonElement.class.isAssignableFrom(property.getType())) {
                property.setValue(ret, element);
            }
        }
        
        return ret;
    }
}

File: src/main/scala/com/lessmarkup/framework/helpers/JsonSerializer.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.framework.helpers

import java.time.OffsetDateTime
import java.util.OptionalLong

import com.google.gson.{JsonElement, JsonNull, JsonObject, JsonParser, JsonPrimitive}

object JsonSerializer {
  def deserializePojo[T <: AnyRef](`type`: Class[T], data: String): Option[T] = {
    deserializePojo(`type`, deserializeToTree(data))
  }

  def serializePojoToTree(`object`: AnyRef): JsonElement = {
    val `type`: Class[_] = `object`.getClass
    if (`type` == classOf[Long]) {
      return new JsonPrimitive(`object`.asInstanceOf[Long])
    }
    else if (`type` == classOf[Int]) {
      return new JsonPrimitive(`object`.asInstanceOf[Int])
    }
    else if (`type` == classOf[String]) {
      return new JsonPrimitive(`object`.asInstanceOf[String])
    }
    else if (classOf[JsonElement].isAssignableFrom(`type`)) {
      return `object`.asInstanceOf[JsonElement]
    }
    val ret: JsonObject = new JsonObject

    for (property <- TypeHelper.getProperties(`object`.getClass)) {
      if (property.getType == classOf[Long]) {
        ret.addProperty(property.getName, property.getValue(`object`).asInstanceOf[Long])
      }
      else if (property.getType == classOf[String]) {
        ret.addProperty(property.getName, property.getValue(`object`).asInstanceOf[String])
      }
      else if (property.getType == classOf[Boolean]) {
        ret.addProperty(property.getName, property.getValue(`object`).asInstanceOf[Boolean])
      }
      else if (property.getType == classOf[Int]) {
        ret.addProperty(property.getName, property.getValue(`object`).asInstanceOf[Int])
      }
      else if (property.getType == classOf[OptionalLong]) {
        val value: OptionalLong = property.getValue(`object`).asInstanceOf[OptionalLong]
        if (value.isPresent) {
          ret.addProperty(property.getName, value.getAsLong)
        }
        else {
          ret.add(property.getName, JsonNull.INSTANCE)
        }
      }
      else if (property.getType == classOf[OffsetDateTime]) {
        ret.addProperty(property.getName, property.getValue(`object`).asInstanceOf[OffsetDateTime].toString)
      }
      else if (classOf[JsonElement].isAssignableFrom(`type`)) {
        ret.add(property.getName, property.getValue(`object`).asInstanceOf[JsonElement])
      }
    }
    ret
  }

  def serializePojo(`object`: AnyRef): String = {
    val obj: JsonElement = serializePojoToTree(`object`)
    obj.toString
  }

  def deserializeToTree(data: String): JsonElement = {
    new JsonParser().parse(data)
  }

  def deserializePojo[T <: AnyRef](dataType: Class[T], data: JsonElement): Option[T] = {

    if (dataType == classOf[Int]) {
      return Option(data.getAsInt.asInstanceOf[Integer].asInstanceOf[T])
    }
    else if (dataType == classOf[Long]) {
      return Option(data.getAsLong.asInstanceOf[T])
    }
    else if (dataType == classOf[String]) {
      return Option(data.getAsString.asInstanceOf[T])
    }
    else if (dataType == classOf[OptionalLong]) {
      if (data.isJsonNull) {
        return Option(OptionalLong.empty.asInstanceOf[T])
      }
      else {
        return Option(OptionalLong.of(data.getAsLong).asInstanceOf[T])
      }
    }
    else if (classOf[JsonElement].isAssignableFrom(dataType)) {
      return Option(data.asInstanceOf[T])
    }

    if (data.isJsonNull || !data.isJsonObject) {
      return None
    }

    val jsonObject: JsonObject = data.getAsJsonObject
    val ret = dataType.newInstance

    TypeHelper.getProperties(dataType)
      .map(p => (p, jsonObject.get(p.getName)))
      .filter(p => p._2 != null)
      .foreach {
      case (property, element) =>
        if (property.getType == classOf[Long]) {
          property.setValue(ret, element.getAsLong.asInstanceOf[AnyRef])
        }
        else if (property.getType == classOf[String]) {
          property.setValue(ret, element.getAsString)
        }
        else if (property.getType == classOf[Boolean]) {
          property.setValue(ret, element.getAsBoolean.asInstanceOf[AnyRef])
        }
        else if (property.getType == classOf[Int]) {
          property.setValue(ret, element.getAsInt.asInstanceOf[AnyRef])
        }
        else if (property.getType == classOf[OptionalLong]) {
          if (!element.isJsonNull) {
            property.setValue(ret, element.getAsLong.asInstanceOf[AnyRef])
          }
        }
        else if (property.getType == classOf[OffsetDateTime]) {
          if (!element.isJsonNull) {
            property.setValue(ret, OffsetDateTime.parse(element.getAsString))
          }
        }
        else if (classOf[JsonElement].isAssignableFrom(property.getType)) {
          property.setValue(ret, element)
        }
    }

    Option(ret)
  }
}